src/pykeen/nn/emb.py

```diff
@@ -0,0 +1,173 @@
+            num_embeddings=num_embeddings,
+            embedding_dim=embedding_dim,
+        )
+
+    @classmethod
+    def init_with_device(
+        cls,
+        num_embeddings: int,
+        embedding_dim: int,
+        device: torch.device,
+        initializer: Optional[Initializer] = None,
+        initializer_kwargs: Optional[Mapping[str, Any]] = None,
+        normalizer: Optional[Normalizer] = None,
+        normalizer_kwargs: Optional[Mapping[str, Any]] = None,
+        constrainer: Optional[Constrainer] = None,
+        constrainer_kwargs: Optional[Mapping[str, Any]] = None,
+    ) -> 'Embedding':  # noqa:E501
+        """Create an embedding object on a device.
+
+        This method is a hotfix for not being able to pass a device during initialization of
+        :class:torch.nn.Embedding`. Instead the weight is always initialized on CPU and has
+        to be moved to GPU afterwards.
+
+        .. seealso::
+
+            https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-37-efficient-random-number-generation-and-application
+
+        :param num_embeddings: >0
+            The number of embeddings.
+        :param embedding_dim: >0
+            The embedding dimensionality.
+        :param device:
+            The device.
+        :param initializer:
+            An optional initializer, which takes a (num_embeddings, embedding_dim) tensor as input, and modifies
+            the weights in-place.
+        :param initializer_kwargs:
+            Additional keyword arguments passed to the initializer
+        :param normalizer:
+            A normalization function
+        :param constrainer:
+            A contrainer applied after each parameter update, without tracking gradients.
+
+        :return:
+            The embedding.
+        """
+        return cls(
+            num_embeddings=num_embeddings,
+            embedding_dim=embedding_dim,
+            initializer=initializer,
+            initializer_kwargs=initializer_kwargs,
+            normalizer=normalizer,
+            normalizer_kwargs=normalizer_kwargs,
+            constrainer=constrainer,
+            constrainer_kwargs=constrainer_kwargs,
+        ).to(device=device)
+
+    @property
+    def num_embeddings(self) -> int:  # noqa: D401
+        """The total number of representations (i.e. the maximum ID)."""
+        return self._embeddings.num_embeddings
+
+    @property
+    def embedding_dim(self) -> int:  # noqa: D401
+        """The representation dimension."""
+        return self._embeddings.embedding_dim
+
+    def reset_parameters(self) -> None:  # noqa: D102
+        # initialize weights in-place
+        self._embeddings.weight.data = self.initializer(self._embeddings.weight.data)
+
+    def post_parameter_update(self):  # noqa: D102
+        # apply constraints in-place
+        if self.constrainer is not None:
+            self._embeddings.weight.data = self.constrainer(self._embeddings.weight.data)
```

Comment:
why is the constrainer an out-of-place function while all of the other ones are in-place functions?

Reply:
It allows using the built-in methods like functional.normalize without having to create in-place operations, which effectively would do the same,  `tensor.data = constrain(tensor.data)`)

Reply:
That makes sense - now looking at the code this is pretty obvious. Was it the other way before?